data/modules/FlightLog/FlightLogEntries.lua

```diff
@@ -0,0 +1,397 @@
+---@field protected sort_date number Seconds since the epoch in game time, used to sort when loading old style
+---@field protected entry string User entered text associated with the entry
+---@field protected always_custom boolean|nil Is this always treated as a custom entry (so not auto deleted)
+FlightLogEntry.Base = utils.class("FlightLogEntry")
+
+---@return string Description of this type
+function FlightLogEntry.Base:GetType()
+	local full_type = self.Class().meta.class
+	return string.sub( full_type, #"FlightLogEntry."+1, #full_type )
+end
+
+---@return boolean true if this is considered to have an entry
+function FlightLogEntry.Base:CanHaveEntry()
+	return true
+end
+
+---@return boolean true if this has an entry
+function FlightLogEntry.Base:HasEntry()
+	return self.entry and #self.entry > 0
+end
+
+---@return string The user provided entry or an empty string if there isn't one.
+function FlightLogEntry.Base:GetEntry()
+	if not self.entry then return "" end
+	return self.entry
+end
+
+---@return boolean true if this has a Delete() method
+function FlightLogEntry.Base:SupportsDelete()
+	return false
+end
+
+
+---@return boolean True if this log entry should be considered custom and so not auto deleted
+function FlightLogEntry.Base:IsCustom()
+	if self.always_custom then return true end
+
+	if not self.entry then return false end
+	if self.entry == "" then return false end
+	return true
+end
+
+-- The serialization table is an array
+-- the elements of the array are a key-value pair
+-- the key is the serialization index (so the type this log entry is)
+-- the value is an array of data, to construct the table from.
+---@param out table<integer, any[]>
+---@return nil
+function FlightLogEntry.Base:AddToSerializationTable( out )
+	local v = {}
+	v[self.GetSerializationIndex()] = self:GetSerializationElements()
+	table.insert(out, v)
+end
+
+---@return integer A unique integer for this specific type
+function FlightLogEntry.Base.GetSerializationIndex()
+	return -1
+end
+
+---@return any[] An array of the elements as they will be serialized
+function FlightLogEntry.Base:GetSerializationElements()
+	return {}
+end
+
+
+---@return string The name for this log entry type
+function FlightLogEntry.Base:GetLocalizedName()
+	return "Error"
+end
+
+---@param earliest_first boolean set to true if your sort order is to show the earlist first dates
+---@return table<string, string>[] An array of key value pairs, the key being localized and the value being formatted appropriately.
+function FlightLogEntry.Base:GetDataPairs( earliest_first )
+	return { { "ERROR", "This should never be seen" } }
+end
+
+---@param entry string A user provided description of the event.
+---If non nil/empty this will cause the entry to be considered custom and not automatically deleted
+---@return nil
+function FlightLogEntry.Base:UpdateEntry( entry )
+
+	if self:IsCustom() then
+		FlightLogEntry.TotalDefaultElements = FlightLogEntry.TotalDefaultElements-1
+	end
```

Comment:
Note: this seems conceptually messy. I don't think it's the responsibility of the individual entries to track how many entries in the overall flightlog are non-custom, and it means there's quite a bit of code scattered around manipulating this variable.

Instead, I'd personally implement it so that adding and updating entries in the flight log goes through the main FlightLog interface and it's the responsibility of those two API functions (e.g. `FlightLog.AddEntryToLog(entry)` / `FlightLog.UpdateEntryNote(entry, note)`) to track changes to the `TotalDefaultElements` count. This also means that constructing a FlightLogEntry variable and then *not* adding it to the FlightLog doesn't wind up corrupting the overall count.

This is maybe not the traditional Object Oriented way to do it, where you get a reference to the entry itself and then ignore the existence of the owning container afterwards, but I think it's a slightly cleaner API model with significantly less internal bookkeeping required.